### ISSUE TITLE
[Feature] Pluggable and enforced Authorization (ideating)

### ISSUE THREAD

**Issue by krishna-ggk** [opensearch-project/security#4702]
(This is early ideation to invite some early thinking/feedback - will evolve this into detailed RFC).

**Problem Statement**

Today security plugin authorization is primarily modeled around transport actions. In addition to transport actions, plugins can directly access shard, cluster-state via listeners and other ways which can be misleading. To achieve zero-trust security posture,
all access to data and metadata need to be encapsulated with authorization enforced.

Further, the authorization rules today are very specific to OpenSearch. There is an opportunity to adopt more open standard models such as [OPA](https://www.openpolicyagent.org/docs/latest/policy-language/), [Cedar](https://www.cedarpolicy.com/en) which offer richer constructs and some of them ensure correctness through formal verification which are potentially good foundations to evolve support on.

**Proposal**

In current architecture, authorization abstraction is implemented in PrivilegesEvaluator invoked through SecurityFilter. The thought process is, can we extract this into a pluggable interface (EvaluationBackend) accepting resource, identity principals, context. The implementations can invoke respective libraries to evaluate permissions as per configured policy.

The enforcement is open and complex item though and will require enlisting all access possibilities.

**Comment by cwperks** [comment]
[Triage] Removing the untriaged label. I like the idea in general and I'd like to see some more concrete examples of what the configuration could look like using Cedar, Opa or Opal. Thank you for filing this issue @krishna-ggk!

**Comment by varun-lodaya** [comment]
Great idea, also looks like a natural extension of security plugin to suit newer and dynamic authorization mechanisms along with existing. How does this fit with the existing work on Resource Based permissions - https://github.com/opensearch-project/security/issues/4500?

**Comment by nibix** [comment]
Just a random thought: Would it make sense to split up the security plugin into one authentication plugin and a separate authorization plugin?

**Comment by nibix** [comment]
@shikharj05 

I believe for authentication and authorization, the code structure already now has quite well visible boundaries, along which a division into two plugins would be relatively easy. The authorization plugin would likely have then a dependency on the authentication plugin (both for logical reasons and for code sharing reasons, e.g. access to the `User` object). 

At least for action based authorization, IMHO, there is no need to introduce new interfaces, as the interface `ActionFilter` can be already considered as the main authorization/access control filter. 

For DLS/FLS, things might be a bit more complicated. But, currently, a contributing factor to this is that the DLS/FLS code is quite badly structured. It would profit from some reorganization anyway. This way, one could make it also more modular.

Regarding audit logging: 

I can imagine audit logging in a separate plugin quite well. One would then have other plugins depending on the audit logging plugin and using the log methods?  However, one could debate whether the audit logging interface should be located in core. A separate plugin would only provide implementations. This would simplify the dependency structure significantly.

Regarding TLS:

Hmmm ... originally, TLS was a separate plugin; the class `OpenSearchSecuritySSLPlugin` is still an artifact of this structure. If it is moved to core, one would need to discuss which parts shall be moved to core. The TLS implementation also has logic to authenticate nodes which want to join the cluster. This might be more security plugin related than core related.

Bottom line: All this is possible, but I actually believe that separating authencation and authorization is the most straight forward possibility.

**Comment by shikharj05** [comment]
> I believe for authentication and authorization, the code structure already now has quite well visible boundaries, along which a division into two plugins would be relatively easy. The authorization plugin would likely have then a dependency on the authentication plugin (both for logical reasons and for code sharing reasons, e.g. access to the User object).

This makes sense to me as well.

> At least for action based authorization, IMHO, there is no need to introduce new interfaces, as the interface ActionFilter can be already considered as the main authorization/access control filter.

While ActionFilter does support it, is there a chance a different plugin implementing an ActionFilter bypasses security (unintentionally) due to ordering of ActionFilters? This was one of the thoughts towards a specific AuthZ interface to ensure no overrides cause unintended results. 

> I can imagine audit logging in a separate plugin quite well. One would then have other plugins depending on the audit logging plugin and using the log methods? However, one could debate whether the audit logging interface should be located in core. A separate plugin would only provide implementations. This would simplify the dependency structure significantly.

Moving it to core or an interface in core is also a good way to think about this. This offers functionality to extend and write custom sinks, etc.

**Comment by nibix** [comment]
> While ActionFilter does support it, is there a chance a different plugin implementing an ActionFilter bypasses security (unintentionally) due to ordering of ActionFilters?

I'd have to double check this in the code, but I *believe* that this is not possible, as action filters do not a provide a possibility to skip over other action filters. They only have means to abort action processing or to modify the request. The ActionFilter from the security plugin uses the lowest possible priority to make sure all potential modifications to the request are already done. One might want to check whether it is possible to put constraints on this priority.

**Comment by cwperks** [comment]
@krishna-ggk FYI Here is an example using Cedar policies for defining access: https://github.com/JanssenProject/jans/wiki/OpenSearch-Cedarling-demo-plugin-v2
